n-api,doc: add info about building n-api addons (#30032)
nodejs/node: doc/api/n-api.md

and by checking, for all external libraries that it uses, that the external
library makes ABI stability guarantees similar to N-API.

## Building

> Unlike modules written in JavaScript, developing and deploying Node native

Reviewer: ```suggestion
Unlike modules written in JavaScript, developing and deploying Node.js native
```

Author: Fixed.

addons using N-API requires an additional set of tools. Besides the basic
tools required to develop for Node.js, the native addon developer requires a
toolchain that can compile C and C++ code into a binary. In addition,
depending upon how the native addon is deployed, the *user* of the native
addon will also need to have a C/C++ toolchain installed.

For Linux developers, the necessary C/C++ toolchain packages are readily
available. For many developers, the [LLVM](https://llvm.org) compiler
infrastructure is a good choice.

For Mac developers, [Xcode](https://developer.apple.com/xcode/) offers all
the required compiler tools. However, it is not necessary to install the
entire Xcode IDE. The following command installs the necessary toolchain:

```bash
xcode-select --install
```

For Windows developers, [Visual Studio](https://visualstudio.microsoft.com)
offers all the required compiler tools. However, it is not necessary to
install the entire Visual Studio IDE. The following command installs the
necessary toolchain:

```bash
npm install --global --production windows-build-tools
```

The sections below describe the additional tools available for developing

> and deploying Node native addons.

Reviewer: ```suggestion
and deploying Node.js native addons.
```

Author: Fixed.

### Build tools

> Both the tools listed here require that *users* of the native
> addon have a C/C++ toolchain installed in order to successfully install

> > the native addon.

Reviewer 2: Why are these lines quoted?

Author: I was attempting to call the reader's attention to this point. But I see the technique is not used anywhere else in the document and I've removed the quote.

system based on [CMake](https://cmake.org).

CMake.js is a good choice for projects that already use CMake or for
developers affected by limitations in node-gyp.

### Uploading precompiled binaries

The two tools listed here permit native addon developers and maintainers to
create and upload binaries to public or private servers. These tools are
typically integrated with CI/CD build systems like
[Travis CI](https://travis-ci.org) and [AppVeyor](https://www.appveyor.com)
to build and upload binaries for a variety of platforms and architectures.
These binaries are then available for download by users who do not need to
have a C/C++ toolchain installed.

#### node-pre-gyp

[node-pre-gyp](https://github.com/mapbox/node-pre-gyp) is a tool based on
node-gyp that adds the ability to upload binaries to a server of the
developer's choice. node-pre-gyp has particularly good support for
uploading binaries to Amazon S3.

#### prebuild

[prebuild](https://github.com/prebuild/prebuild) is a tool that supports
builds using either node-gyp or CMake.js. Unlike node-pre-gyp which
supports a variety of servers, prebuild uploads binaries only to

> [GitHub releases](https://github.com/prebuild/prebuild). prebuild is a

Reviewer 2: Did you intend that "GitHub releases" point to the same thing as "prebuild"?

Author: Fixed.

good choice for GitHub projects using CMake.js.

> 

Reviewer: I'd also mention https://github.com/prebuild/prebuildify. Shipping the binaries in the npm package is the best approach imho.

Author: I've added a section for prebuildify. Thank you for the tip.

and by checking, for all external libraries that it uses, that the external
library makes ABI stability guarantees similar to N-API.

## Building

> Unlike addons coded in JavaScript, developing and deploying Node native

Reviewer 3: ```suggestion
Unlike modules written in JavaScript, developing and deploying Node.js native
```

Author: Changed.

addons using N-API requires an additional set of tools. Besides the basic

> tools required to develop for Node, the native addon developer requires a

Reviewer 3: ```suggestion
tools required to develop for Node.js, the native addon developer requires a
```

Author: Changed.

toolchain that can compile C and C++ code into a binary. In addition,
depending upon how the native addon is deployed, the *user* of the native
addon will also need to have a C/C++ toolchain installed.

For Linux developers, the necessary C/C++ toolchain is typically already

> installed. For systems lacking the necessary toolchain, the

Reviewer 3: Maybe it’s just my Ubuntu experience but installing them manually is part of my standard laptop setup routine – I would maybe use wording that implies that packages typically exist, rather than that they are already installed to begin with.

Author: Changed.

entire Xcode IDE. The following command installs the necessary toolchain:

```bash
xcode-select --install
```

For Windows developers, [Visual Studio](https://visualstudio.microsoft.com)
offers all the required compiler tools. However, it is not necessary to
install the entire Visual Studio IDE. The following command installs the
necessary toolchain:

```bash
npm install --global --production windows-build-tools
```

The sections below describe the additional tools available for developing
and deploying Node native addons.

### Build tools

> Both the tools listed here require that *users* of the native
> addon have a C/C++ toolchain installed in order to successfully install
> the native addon.

#### node-gyp

[node-gyp](https://github.com/nodejs/node-gyp) is a build system based on
Google's [GYP](https://gyp.gsrc.io) tool and comes bundled with npm. GYP,
and therefore node-gyp, requires that Python be installed.

Historically, node-gyp has been the tool of choice for building native
addons. However, Google has stopped development on GYP and some developers
are concerned about the dependency of some GYP versions on Python 2.7 which

> has a sunset date of 2020.

Reviewer 3: I don’t think the latter issue is a concern anymore if you’re using the latest versions of npm/node-gyp, so I’d omit both that and the statement that Google has stopped development – it’s not like we’re in a position to actually abandon support for node-gyp.

What you *could* mention is that developers have found node-gyp to have limitations. It’s somewhat opinionated, but I am sure many of us would agree with such a statement.

Author: Changed.

and by checking, for all external libraries that it uses, that the external
library makes ABI stability guarantees similar to N-API.

## Building

Unlike modules written in JavaScript, developing and deploying Node.js native
addons using N-API requires an additional set of tools. Besides the basic
tools required to develop for Node.js, the native addon developer requires a
toolchain that can compile C and C++ code into a binary. In addition,
depending upon how the native addon is deployed, the *user* of the native
addon will also need to have a C/C++ toolchain installed.

For Linux developers, the necessary C/C++ toolchain packages are readily

> available. For many developers, the [LLVM](https://llvm.org) compiler

Reviewer 4: I'm wondering if we should reference GCC instead since I think that is what we use in the community build/testing on many platforms.

Author: I've added GCC as the primary recommendation in addition to LLVM as an alternative.

entire Xcode IDE. The following command installs the necessary toolchain:

```bash
xcode-select --install
```

For Windows developers, [Visual Studio](https://visualstudio.microsoft.com)
offers all the required compiler tools. However, it is not necessary to
install the entire Visual Studio IDE. The following command installs the
necessary toolchain:

```bash
npm install --global --production windows-build-tools
```

The sections below describe the additional tools available for developing
and deploying Node.js native addons.

### Build tools

Both the tools listed here require that *users* of the native
addon have a C/C++ toolchain installed in order to successfully install
the native addon.

#### node-gyp

[node-gyp](https://github.com/nodejs/node-gyp) is a build system based on
Google's [GYP](https://gyp.gsrc.io) tool and comes bundled with npm. GYP,
and therefore node-gyp, requires that Python be installed.

Historically, node-gyp has been the tool of choice for building native
addons. It has widespread adoption and documentation. However, some
developers have run into limitations in node-gyp.

#### CMake.js

> [CMake.js](https://github.com/cmake-js/cmake-js) is an alternative build

Reviewer 2: nit: there is a mixture of inline links such as these and bottom-of-document links in this document. We should use only bottom-of-document links – but I'm OK if we move that out of scope for this PR.

Author: I've move all links in the document to be end notes.
